### ISSUE TITLE
Use different input_sizes on Densenet121 Classifier

### ISSUE THREAD

**Issue by JuanPabloMontoya271** [Project-MONAI/MONAI#695]
Hi, I have a CT dataset, and I am trying to train as in the MR classification example, the thing is, my CTs have a volume dimension of arround 36 and 46, and in the example, they use a resize of 96,96,96 shape, which is not optimal for my input shape. which is originaly (512,512, ~40), the thing is that I dont know how to select the correct input dimensions in order to make the densenet model to work. I tried to modify only the volume dimension to 48, which is divisible of 96, and it gave me an error, that said:

expected stack dimension to be the same, found [10, 1,48,96,96] on stack 1, and [10, 1,96,96,48] on stack 3

**Comment by Nic-Ma** [comment]
Hi @JuanPabloMontoya271 ,

Thanks for your interest here!
1. I think you can use the `Resize` or `Resized` transform, and set the `spatial_size` to (-1, -1, 48) to only resize the `D` dim.
2. Could you please help share your test code? I didn't understand why your images have different shape: [10, 1, 48, 96, 96] and [10, 1, 96, 96, 48].

Thanks.

**Comment by JuanPabloMontoya271** [comment]
def main():

    
    train_files = train
    val_files = test
    train_transforms = Compose(
        [
            LoadNiftid(keys=["img"]),
            AddChanneld(keys=["img"]),
            ScaleIntensityd(keys=["img"]),
            Resized(keys=["img"], spatial_size=(96,96,48 )),
            RandRotate90d(keys=["img"], prob=0.8, spatial_axes=[0, 2]),
            ToTensord(keys=["img"]),
        ]
    )
    val_transforms = Compose(
        [
            LoadNiftid(keys=["img"]),
            AddChanneld(keys=["img"]),
            ScaleIntensityd(keys=["img"]),
            Resized(keys=["img"], spatial_size=( 96,96,48)),
            ToTensord(keys=["img"]),
        ]
    )

    check_ds = monai.data.Dataset(data=train_files, transform=train_transforms)
    check_loader = DataLoader(check_ds, batch_size=2, num_workers=4, pin_memory=torch.cuda.is_available())
    check_data = monai.utils.misc.first(check_loader)
    print(check_data["img"].shape, check_data["label"])
    train_ds = monai.data.Dataset(data=train_files, transform=train_transforms)
    train_loader = DataLoader(train_ds, batch_size=10, shuffle=True, num_workers=4, pin_memory=torch.cuda.is_available())

    # create a validation data loader
    val_ds = monai.data.Dataset(data=val_files, transform=val_transforms)
    val_loader = DataLoader(val_ds, batch_size=10, num_workers=4, pin_memory=torch.cuda.is_available())

    # Create DenseNet121, CrossEntropyLoss and Adam optimizer
    device = torch.device("cuda:0")
    model = monai.networks.nets.densenet.densenet169(spatial_dims=3, in_channels=1, out_channels=2).to(device)
    loss_function = torch.nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(model.parameters(), 1e-5)

    # start a typical PyTorch training
    val_interval = 2
    best_metric = -1
    best_metric_epoch = -1
    writer = SummaryWriter()
    epochs = 50
    aucs = []
    losses=[]
    accs =[]
    for epoch in range(epochs):
        print("-" * 10)
        print(f"epoch {epoch + 1}/{epochs}")
        model.train()
        epoch_loss = 0
        step = 0
      
        for batch_data in train_loader:
            step += 1
            inputs, labels = batch_data["img"].to(device), batch_data["label"].to(device)
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = loss_function(outputs, labels)
            loss.backward()
            optimizer.step()
            epoch_loss += loss.item()
            epoch_len = len(train_ds) // train_loader.batch_size
            print(f"{step}/{epoch_len}, train_loss: {loss.item():.4f}")
            writer.add_scalar("train_loss", loss.item(), epoch_len * epoch + step)
        epoch_loss /= step
        print(f"epoch {epoch + 1} average loss: {epoch_loss:.4f}")
        losses.append(epoch_loss)
        if (epoch + 1) % val_interval == 0:
            model.eval()
            with torch.no_grad():
                y_pred = torch.tensor([], dtype=torch.float32, device=device)
                y = torch.tensor([], dtype=torch.long, device=device)
                for val_data in val_loader:
                    val_images, val_labels = val_data["img"].to(device), val_data["label"].to(device)
                    y_pred = torch.cat([y_pred, model(val_images)], dim=0)
                    y = torch.cat([y, val_labels], dim=0)

                acc_value = torch.eq(y_pred.argmax(dim=1), y)
                acc_metric = acc_value.sum().item() / len(acc_value)
                auc_metric = compute_roc_auc(y_pred, y, to_onehot_y=True)
                if acc_metric > best_metric:
                    best_metric = acc_metric
                    best_metric_epoch = epoch + 1
                    torch.save(model.state_dict(), "best_metric_model.pth")
                    print("saved new best metric model")
                print(
                    "current epoch: {} current accuracy: {:.4f} current AUC: {:.4f} best accuracy: {:.4f} at epoch {}".format(
                        epoch + 1, acc_metric, auc_metric, best_metric, best_metric_epoch
                    )
                )
                accs.append(acc_metric)
                aucs.append(auc_metric)
                writer.add_scalar("val_accuracy", acc_metric, epoch + 1)
    print(f"train completed, best_metric: {best_metric:.4f} at epoch: {best_metric_epoch}")
    writer.close()
    return aucs, losses, accs

**Comment by JuanPabloMontoya271** [comment]
this is the code, it is the same as the 3D classification example 
I just changed the input size on resized

**Comment by Nic-Ma** [comment]
Hi @JuanPabloMontoya271 ,

As your images have different size at H, W, D dims, I think you should Resize after random rotate, which is the root cause of your error:
```py
train_transforms = Compose(
    [
        LoadNiftid(keys=["img"]),
        AddChanneld(keys=["img"]),
        ScaleIntensityd(keys=["img"]),
        Resized(keys=["img"], spatial_size=(96,96,48 )),
        RandRotate90d(keys=["img"], prob=0.8, spatial_axes=[0, 2]),
        ToTensord(keys=["img"]),
    ]
)
```
And I am not sure whether you need `random rotate` for your dataset as augmentation, please double confirm that.

Thanks.

**Comment by JuanPabloMontoya271** [comment]
I will remove it, so to confirm, I can choose any input size, and it will work with the neural net?
